xds/internal/balancer/outlierdetection/subconn_wrapper.go

```diff
@@ -31,23 +33,43 @@ import (
 // whether or not this SubConn is ejected.
 type subConnWrapper struct {
 	balancer.SubConn
+	// The following fields are set during object creation and read-only after
+	// that.
+
 	listener func(balancer.SubConnState)
 
 	// addressInfo is a pointer to the subConnWrapper's corresponding address
 	// map entry, if the map entry exists.
 	addressInfo unsafe.Pointer // *addressInfo
 	// These two pieces of state will reach eventual consistency due to sync in
 	// run(), and child will always have the correctly updated SubConnState.
-	// latestState is the latest state update from the underlying SubConn. This
-	// is used whenever a SubConn gets unejected.
-	latestState balancer.SubConnState
-	ejected     bool
+	// latestDeleveredState is the latest state update from the underlying SubConn. This
+	// is used whenever a SubConn gets unejected. This will be the health state
+	// if a health listener is being used, otherwise it will be the connectivity
+	// state.
+	latestDeleveredState balancer.SubConnState
+	ejected              bool
 
 	scUpdateCh *buffer.Unbounded
 
 	// addresses is the list of address(es) this SubConn was created with to
 	// help support any change in address(es)
 	addresses []resolver.Address
+	// healthListenerEnabled indicates whether the leaf LB policy is using a
+	// generic health listener. When enabled, ejection updates are sent via the
+	// health listener instead of the connectivity listener. Once Dualstack
+	// changes are complete, all SubConns will be created by pickfirst which
+	// uses the health listener.
+	healthListenerEnabled bool
+
+	// Access to the following fields are protected by a mutex.
+	mu             sync.Mutex
```

Comment:
With the introduction of this mutex, we now have three mutexes within this LB policy and I feel that it is one too many. One option that I can think of simplify things is as follows:
- have this mutex here, but do not grab it from outside this type. Instead write methods to get/set fields guarded by this mutex. So, the LB policy will no longer directly access fields here (with or without the mutex), but will instead call methods on this type.
- the `childMu` mutex on the LB policy is always only used when invoking methods on the child policy. What if we make a wrapper type `childPolicyWrapper`, that contains this mutex and the child policy. Again, this will free up the LB policy from having to acquire this mutex every time it has to call any methods on the child policy. Instead the `childPolicyWrapper` will provide methods around the LB policy API that simply grab the lock and release it with a defer and call the underlying method on the child policy.

Let me know what you think.